src/engine/threaded_engine.cc

```diff
@@ -309,6 +309,13 @@ void ThreadedEngine::PushAsync(AsyncFn fn, Context exec_ctx,
                                int priority,
                                const char* opr_name,
                                bool wait) {
+#if MXNET_USE_CUDA
+  if (exec_ctx.dev_mask() == gpu::kDevMask) {
+    int device_count = 0;
+    cudaGetDeviceCount(&device_count);
+    CHECK_LT(exec_ctx.dev_id, device_count) << "Incorrect GPU device id";
```

Comment:
error message should be more descriptive. i.e. what the id is and why its illegal